service/stream/exec_data_streamer.go

```diff
@@ -0,0 +1,168 @@
+	cfg := DefaultConfig
+	for _, option := range options {
+		option(&cfg)
+	}
+
+	ctx := context.Background()
+
+	// initialize clients
+	opts := []grpc.DialOption{grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(grpcutils.DefaultMaxMsgSize)),
+		grpc.WithTransportCredentials(insecure.NewCredentials())}
+	conn, err := grpc.Dial(accessAddr, opts...)
+	if err != nil {
+		log.Error()
+	}
+	execDataAPI := execData.NewExecutionDataAPIClient(conn)
+	accessAPI := access.NewAccessAPIClient(conn)
+
+	// get chainID from network params
+	params, err := accessAPI.GetNetworkParameters(ctx, &access.GetNetworkParametersRequest{})
+	if err != nil {
+
+	}
+	chain := flow.ChainID(params.ChainId)
+
+	return &ExecDataStreamer{
+		log:       log,
+		execApi:   execDataAPI,
+		accessApi: accessAPI,
+		queue:     archive.NewDeque(),
+		buffer:    archive.NewDeque(),
+		limit:     cfg.BufferSize,
+		busy:      0,
+		ctx:       ctx,
+		chain:     chain,
+	}
+}
+
+// OnBlockFinalized is a callback for the Flow consensus follower. It is called
+// each time a block is finalized by the Flow consensus algorithm.
+func (e *ExecDataStreamer) OnBlockFinalized(block *model.Block) {
+	blockID := block.BlockID
+	// We push the block ID to the front of the queue; the streamer will try to
+	// download the blocks in a FIFO manner.
+	e.queue.PushFront(blockID)
+
+	e.log.Debug().Hex("block", blockID[:]).Msg("execution record queued for download")
+}
+
+// Next returns the next available block data. It returns an ErrUnavailable if no block
+// data is available at the moment.
+func (e *ExecDataStreamer) Next() (*uploader.BlockData, error) {
+	// same implementation as GCPStreamer
+	go e.poll()
+	if e.buffer.Len() == 0 {
+		e.log.Debug().Msg("buffer empty, no execution record available")
+		return nil, archive.ErrUnavailable
+	}
+	record := e.buffer.PopBack()
+	return record.(*uploader.BlockData), nil
+}
+
+func (e *ExecDataStreamer) poll() {
+	// same implementation as GCPStreamer
+	if !atomic.CompareAndSwapUint32(&e.busy, 0, 1) {
+		return
+	}
+	defer atomic.StoreUint32(&e.busy, 0)
+	err := e.pullExecData()
+	if errors.Is(err, storage.ErrObjectNotExist) {
+		e.log.Debug().Msg("next execution record not available, download stopped")
+		return
+	}
+	if err != nil {
+		e.log.Error().Err(err).Msg("could not download execution records")
+		return
+	}
+}
+
+func (e ExecDataStreamer) pullExecData() error {
+	for {
+		// same implementation as GCPStreamer
+		if uint(e.buffer.Len()) >= e.limit {
+			e.log.Debug().Uint("limit", e.limit).Msg("buffer full, stopping execution record download")
+			return nil
+		}
+
+		if uint(e.queue.Len()) == 0 {
+			e.log.Debug().Msg("queue empty, stopping execution record download")
+			return nil
+		}
+		blockID := e.queue.PopBack().(flow.Identifier)
+		record, err := e.getUploaderBlockData(blockID)
+		if err != nil {
+			e.queue.PushBack(blockID)
+			return fmt.Errorf("could not pull execution record (name: %s): %w", blockID, err)
+		}
+
+		e.log.Debug().
```

Comment:
This log is worth to be INFO level
```suggestion
		e.log.Info().
```